File: JEDI/bin/leadTime.rb
class LeadTime

  ##################################
  #
  # fcstToSeconds
  #
  ##################################
  def self.fcstToSeconds(fcst)

    seconds=0
    if fcst =~ /[^\d]+(\d+)S/
      seconds += $1.to_i
    end
    if fcst =~ /[^\d]+(\d+)M/
      seconds += $1.to_i * 60
    end
    if fcst =~ /[^\d]+(\d+)H/
      seconds += $1.to_i * 3600
    end
    if fcst =~ /[^\d]+(\d+)D/
      seconds += $1.to_i * 3600 * 24
    end
    if fcst =~ /^M/   
      seconds = seconds * -1
    else
      seconds
    end

  end


  ##################################
  #
  # secondsToFcst
  #
  ##################################
  def self.secondsToFcst(s)

    seconds = s.to_i

    fcst = ""
    if seconds < 0 
      fcst = "M"
    else
      fcst = "P"
    end

    days = seconds / (3600 * 24)
    if days > 0
      fcst = fcst + "#{days}D"
    end

    fcst = fcst + "T" unless (seconds % (3600 * 24) == 0) && (seconds != 0)

    hours = (seconds - days * 3600 * 24) / 3600
    if hours > 0
      fcst = fcst + "#{hours}H"
    end

    minutes = (seconds - days * 3600 * 24 - hours * 3600) / 60
    if minutes > 0
      fcst = fcst + "#{minutes}M"
    end

    seconds = seconds - days * 3600 * 24 - hours * 3600 - minutes * 60
    if (days == 0 && hours == 0 && minutes == 0) || (seconds > 0)
      fcst = fcst + "#{seconds}S"
    end

    fcst

  end

end

File: src/chiltepin/jedi/leadtime.py
def leadtime_to_seconds(leadtime):
    import re

    seconds = 0
    p = re.compile(r"\D+(\d+)S")
    m = p.search(leadtime)
    if m:
        seconds += int(m.group(1))
    p = re.compile(r"\D+(\d+)M")
    m = p.search(leadtime)
    if m:
        seconds += int(m.group(1)) * 60
    p = re.compile(r"\D+(\d+)H")
    m = p.search(leadtime)
    if m:
        seconds += int(m.group(1)) * 3600
    p = re.compile(r"\D+(\d+)D")
    m = p.search(leadtime)
    if m:
        seconds += int(m.group(1)) * 3600 * 24
    p = re.compile("M")
    m = p.match(leadtime)
    if m:
        seconds = seconds * -1
    return seconds


def seconds_to_leadtime(s):
    leadtime = ""
    if s < 0:
        leadtime = "M"
        seconds = s * -1
    else:
        leadtime = "P"
        seconds = s
    days = seconds // (3600 * 24)
    if days > 0:
        leadtime = leadtime + f"{days}D"
    if seconds % (3600 * 24) != 0 or seconds == 0:
        leadtime = leadtime + "T"
    hours = (seconds - days * 3600 * 24) // 3600
    if hours > 0:
        leadtime = leadtime + f"{hours}H"
    minutes = (seconds - days * 3600 * 24 - hours * 3600) // 60
    if minutes > 0:
        leadtime = leadtime + f"{minutes}M"
    seconds = seconds - days * 3600 * 24 - hours * 3600 - minutes * 60
    if (days == 0 and hours == 0 and minutes == 0) or (seconds > 0):
        leadtime = leadtime + f"{seconds}S"
    return leadtime
